server/handler.go

```diff
@@ -115,27 +142,34 @@ func (h *Handler) ComQuery(
 
 			r = nil
 			proccesedAtLeastOneBatch = true
-
 			continue
 		}
 
-		row, err := rows.Next()
-		if err != nil {
+		select {
+		case err := <-errchan:
 			if err == io.EOF {
-				break
+				goto EndFor
 			}
 
 			return err
-		}
+		case row := <-rowchan:
+			outputRow, err := rowToSQL(schema, row)
+			if err != nil {
+				return err
```

Comment:
Be careful: Returning at this point will orphan the goroutine that is reading from the rows channel, unless this is the last value in the iterator. The goroutine cannot exit because it will block trying to deliver a value and nothing will read any further values after this return. A goroutine cannot be GC'd until it exits. As a rule of thumb, try to make sure any goroutine you start will either run forever (the lifetime of the program) or has a well-defined way to return even in case of error.

Usually that means you either need to fully drain any channel you create, or ensure that the operation populating the channel is cancelled before exit. One way you might do that here:

```go
rctx, cancel := context.WithCancel(ctx)
defer cancel()
…
// in the goroutine
for {
	…
	select {
	case <-rctx.Done():
		return
	case rowc <- row:
		// ok
	}
```

This is imperfect, since it could still stall trying to write an error: You can work around that by buffering `errchan` with 1 slot.

Reply:
This blocked for me, I used instead a signal channel to exit the goroutine in that case (with a buffer with 1 slot, because it could be busy getting the next row and block the sender), seems to work, I also refactored the logic a little to remote the `return`s from the `case`s which always make me feel uneasy, PTAL.

Reply:
Hmm. It shouldn't block as long as at least one of those options can proceed. Under the covers, the context is really just using a signal channel anyway. What I had in mind was something along these lines: https://play.golang.org/p/LOM4jRDubmv

Anyway, though, I think a separate channel is also fine, but see my note on the other thread.